job_executor/config/log.py

```diff
@@ -1,114 +1,98 @@
+import sys
 import json
 import logging
 import logging.handlers
-import sys
-import traceback
-from datetime import datetime, UTC
-from multiprocessing import Queue
-
-from json_logging import util
+import datetime
 import tomlkit
+import threading
 
+from typing import Tuple
+from multiprocessing import Queue
 from job_executor.config import environment
 
 
-class ContextFilter(logging.Filter):
-    """
-    This is a filter which injects contextual information into the log.
-    """
+def _get_project_meta():
+    with open("pyproject.toml", encoding="utf-8") as pyproject:
+        file_contents = pyproject.read()
+    return tomlkit.parse(file_contents)["tool"]["poetry"]
+
+
+class MicrodataJSONFormatter(logging.Formatter):
+    def __init__(self):
+        self.pkg_meta = _get_project_meta()
+        self.host = environment.get("DOCKER_HOST_NAME")
+        self.command = json.dumps(sys.argv)
+
+    def format(self, record: logging.LogRecord) -> str:
+        return json.dumps(
+            {
+                "@timestamp": datetime.datetime.fromtimestamp(
+                    record.created,
+                    tz=datetime.timezone.utc,
+                ).strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3]
+                + "Z",
+                "command": self.command,
+                "error.stack": record.__dict__.get("exc_info"),
+                "host": self.host,
+                "message": record.getMessage(),
+                "level": record.levelno,
+                "levelName": record.levelname,
+                "loggerName": record.name,
+                "schemaVersion": "v3",
+                "serviceName": "job-executor",
+                "serviceVersion": str(self.pkg_meta["version"]),
+                "thread": record.threadName,
+            }
+        )
+
 
+class WorkerFormatter(logging.Formatter):
     job_id = ""
 
     def __init__(self, job_id: str):
-        super().__init__()
         self.job_id = job_id
 
-    def filter(self, record):
-        # prefix record.msg instead of adding a new field
-        # to be compliant with Kibana
-        record.msg = self.job_id + ": " + str(record.msg)
-        return True
-
-
-def configure_worker_logger(queue: Queue, job_id: str):
-    queue_handler = logging.handlers.QueueHandler(queue)
-    queue_handler.setLevel(logging.INFO)
-    # use default formatter here so that the JSON formatter will be
-    # applied when reading from queue
-    queue_handler.formatter = logging.Formatter()
+    def format(self, record: logging.LogRecord) -> str:
+        return f"{self.job_id}: {record.msg}"
```

Comment:
If the worker returns only the message, and the main thread does the rest of the logging, can the timestamp be a bit off here?

Reply:
Hmm I did not consider this. This actually extends to error stacks also, I'll have a second look 💪🏻